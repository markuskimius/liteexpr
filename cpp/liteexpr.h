#ifndef LITEEXPR_H_
#define LITEEXPR_H_

#include <any>
#include <map>
#include <string>
#include <vector>
#include <memory>
#include <utility>
#include <cstdint>
#include <codecvt>
#include <iostream>
#include <initializer_list>
#include "antlr4-runtime.h"
#include "LiteExprLexer.h"
#include "LiteExprParser.h"
#include "LiteExprBaseVisitor.h"


/* ***************************************************************************
* VALUE
*/

namespace liteexpr {
    using std::any;
    using std::map;
    using std::pair;
    using std::vector;
    using std::string;
    using std::type_info;
    using std::shared_ptr;
    using std::initializer_list;

    class Value;
    class Integer;
    class Double;
    class String;
    class Array;
    class Object;
    class SymbolTable;
    class Ident;
    class Function;
    class Evaluator;
    typedef shared_ptr<Value> VALUE;
    typedef shared_ptr<Integer> INTEGER;
    typedef shared_ptr<Double> DOUBLE;
    typedef shared_ptr<String> STRING;
    typedef shared_ptr<Array> ARRAY;
    typedef shared_ptr<Object> OBJECT;
    typedef shared_ptr<SymbolTable> SYMBOLS;
    typedef shared_ptr<Ident> IDENT;
    typedef shared_ptr<Function> FUNCTION;
    extern int64_t MAXARGS;

    class Value {
        public:
            virtual string name() const=0;
            virtual string encoded() const=0;
            virtual const type_info& type() const=0;

            virtual bool istrue() const;
            virtual int64_t length() const;
            virtual int64_t ivalue() const;
            virtual double dvalue() const;
            virtual string svalue() const;
            virtual const vector<VALUE>& avalue() const;
            virtual const map<string,VALUE>& ovalue() const;
            virtual VALUE op_not() const;
            virtual VALUE op_inv() const;
            virtual VALUE op_pos() const;
            virtual VALUE op_neg() const;
            virtual VALUE op_inc() const;
            virtual VALUE op_dec() const;
            virtual VALUE op_lt(const VALUE other) const;
            virtual VALUE op_gt(const VALUE other) const;
            virtual VALUE op_eq(const VALUE other) const;
            virtual VALUE op_ne(const VALUE other) const;
            virtual VALUE op_lte(const VALUE other) const;
            virtual VALUE op_gte(const VALUE other) const;
            virtual VALUE op_pow(const VALUE other) const;
            virtual VALUE op_mul(const VALUE other) const;
            virtual VALUE op_div(const VALUE other) const;
            virtual VALUE op_add(const VALUE other) const;
            virtual VALUE op_sub(const VALUE other) const;
            virtual VALUE op_mod(const VALUE other) const;
            virtual VALUE op_shl(const VALUE other) const;
            virtual VALUE op_asr(const VALUE other) const;
            virtual VALUE op_shr(const VALUE other) const;
            virtual VALUE op_and(const VALUE other) const;
            virtual VALUE op_xor(const VALUE other) const;
            virtual VALUE op_or(const VALUE other) const;
    };

    class Integer: public Value {
        int64_t value;

        public:
            Integer(int64_t v);
            int64_t native() const;
            static string encode(int64_t decoded);
            static int64_t decode(const string& encoded);
            static int64_t decodeHex(const string& encoded);

            string name() const override;
            string encoded() const override;
            const type_info& type() const override;

            bool istrue() const override;
            int64_t ivalue() const override;
            double dvalue() const override;
            string svalue() const override;
            VALUE op_inv() const override;
            VALUE op_pos() const override;
            VALUE op_neg() const override;
            VALUE op_inc() const override;
            VALUE op_dec() const override;
            VALUE op_lt(const VALUE other) const override;
            VALUE op_gt(const VALUE other) const override;
            VALUE op_eq(const VALUE other) const override;
            VALUE op_ne(const VALUE other) const override;
            VALUE op_lte(const VALUE other) const override;
            VALUE op_gte(const VALUE other) const override;
            VALUE op_pow(const VALUE other) const override;
            VALUE op_mul(const VALUE other) const override;
            VALUE op_div(const VALUE other) const override;
            VALUE op_add(const VALUE other) const override;
            VALUE op_sub(const VALUE other) const override;
            VALUE op_mod(const VALUE other) const override;
            VALUE op_shl(const VALUE other) const override;
            VALUE op_asr(const VALUE other) const override;
            VALUE op_shr(const VALUE other) const override;
            VALUE op_and(const VALUE other) const override;
            VALUE op_xor(const VALUE other) const override;
            VALUE op_or(const VALUE other) const override;
    };

    class Double: public Value {
        double value;

        public:
            Double(double v);
            double native() const;
            static string encode(double decoded);
            static double decode(const string& encoded);

            string name() const override;
            string encoded() const override;
            const type_info& type() const override;

            bool istrue() const override;
            int64_t ivalue() const override;
            double dvalue() const override;
            string svalue() const override;
            VALUE op_pos() const override;
            VALUE op_neg() const override;
            VALUE op_lt(const VALUE other) const override;
            VALUE op_gt(const VALUE other) const override;
            VALUE op_eq(const VALUE other) const override;
            VALUE op_ne(const VALUE other) const override;
            VALUE op_lte(const VALUE other) const override;
            VALUE op_gte(const VALUE other) const override;
            VALUE op_pow(const VALUE other) const override;
            VALUE op_mul(const VALUE other) const override;
            VALUE op_div(const VALUE other) const override;
            VALUE op_add(const VALUE other) const override;
            VALUE op_sub(const VALUE other) const override;
    };

    class String: public Value {
        string value;

        public:
            String(const string& v);
            const string& native() const;
            static string encode(const string& decoded);
            static string decode(const string& encoded);
            static vector<string> split(const string& text, const string& delim);

            string name() const override;
            string encoded() const override;
            const type_info& type() const override;

            bool istrue() const override;
            int64_t length() const override;
            string svalue() const override;
            VALUE op_lt(const VALUE other) const override;
            VALUE op_gt(const VALUE other) const override;
            VALUE op_eq(const VALUE other) const override;
            VALUE op_ne(const VALUE other) const override;
            VALUE op_lte(const VALUE other) const override;
            VALUE op_gte(const VALUE other) const override;
            VALUE op_add(const VALUE other) const override;
    };

    class Array: public Value {
        vector<VALUE> value;

        public:
            Array();
            Array(const vector<VALUE>& v);
            const vector<VALUE>& native() const;
            static string encode(const vector<VALUE>& decoded);
            VALUE get(int64_t i) const;
            void set(int64_t i, VALUE v);
            void push(VALUE v);

            string name() const override;
            string encoded() const override;
            const type_info& type() const override;

            bool istrue() const override;
            int64_t length() const override;
            const vector<VALUE>& avalue() const override;
            string svalue() const override;
            VALUE op_lt(const VALUE other) const override;
            VALUE op_gt(const VALUE other) const override;
            VALUE op_eq(const VALUE other) const override;
            VALUE op_ne(const VALUE other) const override;
            VALUE op_lte(const VALUE other) const override;
            VALUE op_gte(const VALUE other) const override;
            VALUE op_add(const VALUE other) const override;
    };

    class Object: public Value {
        protected:
            map<string,VALUE> value;
            Object();

        public:
            Object(initializer_list<pair<string,VALUE> > init);
            Object(const map<string,VALUE>& v);
            const map<string,VALUE>& native() const;
            static string encode(const map<string,VALUE>& decoded, SYMBOLS parent=nullptr);
            virtual VALUE get(const string& k);
            virtual void set(const string& k, VALUE v);
            virtual bool has(const string& k);

            virtual string name() const override;
            virtual string encoded() const override;
            virtual const type_info& type() const override;

            virtual bool istrue() const override;
            virtual int64_t length() const override;
            virtual const map<string,VALUE>& ovalue() const override;
            virtual string svalue() const override;
            virtual VALUE op_eq(const VALUE other) const override;
            virtual VALUE op_ne(const VALUE other) const override;
    };

    class SymbolTable: public Object {
        SYMBOLS parent;
        SYMBOLS root;

        public:
            SymbolTable(initializer_list<pair<string,VALUE> > init);
            SymbolTable(SYMBOLS parent=nullptr);
            VALUE get(const string& k) override;
            void set(const string& k, VALUE v) override;
            bool has(const string& k) override;

            string encoded() const override;
    };

    class Function: public Value {
        VALUE (*func)(const vector<VALUE>&);
        VALUE (*dfunc)(vector<LiteExprParser::ExprContext*>& vexpr, Evaluator* visitor);
        VALUE (*xfunc)(vector<LiteExprParser::ExprContext*>& vexpr, Evaluator* visitor, SYMBOLS scope);
        LiteExprParser::ExprContext* staticExpr;
        int64_t minargs;
        int64_t maxargs;
        SYMBOLS scope;

        public:
            Function(VALUE (*func)(const vector<VALUE>& vv), int64_t minargs=0, int64_t maxargs=MAXARGS);
            Function(VALUE (*func)(vector<LiteExprParser::ExprContext*>& vexpr, Evaluator* visitor), int64_t minargs=0, int64_t maxargs=MAXARGS);
            Function(VALUE (*func)(vector<LiteExprParser::ExprContext*>& vexpr, Evaluator* visitor, SYMBOLS scope), SYMBOLS scope, int64_t minargs=0, int64_t maxargs=MAXARGS);

            VALUE call(vector<LiteExprParser::ExprContext*>& vexpr, Evaluator* visitor) const;
            SYMBOLS getScope() const;

            string name() const override;
            string encoded() const override;
            const type_info& type() const override;

            void setStaticExpr(LiteExprParser::ExprContext*);
    };
}


/* ***************************************************************************
* VARIABLE
*/

namespace liteexpr {
    class Ident: public Value {
        VALUE container;
        VALUE key;

        public:
            Ident(VALUE container, VALUE key);
            virtual ~Ident();
            void set(VALUE other);
            VALUE get() const;
            VALUE getkey() const;

            string name() const override;
            string encoded() const override;
            const type_info& type() const override;

            bool istrue() const override;
            int64_t length() const override;
            int64_t ivalue() const override;
            double dvalue() const override;
            string svalue() const override;
            const vector<VALUE>& avalue() const override;
            const map<string,VALUE>& ovalue() const override;

            VALUE op_not() const override;
            VALUE op_inv() const override;
            VALUE op_pos() const override;
            VALUE op_neg() const override;
            VALUE op_inc() const override;
            VALUE op_dec() const override;
            VALUE op_lt(const VALUE other) const override;
            VALUE op_gt(const VALUE other) const override;
            VALUE op_eq(const VALUE other) const override;
            VALUE op_ne(const VALUE other) const override;
            VALUE op_lte(const VALUE other) const override;
            VALUE op_gte(const VALUE other) const override;
            VALUE op_pow(const VALUE other) const override;
            VALUE op_mul(const VALUE other) const override;
            VALUE op_div(const VALUE other) const override;
            VALUE op_add(const VALUE other) const override;
            VALUE op_sub(const VALUE other) const override;
            VALUE op_mod(const VALUE other) const override;
            VALUE op_shl(const VALUE other) const override;
            VALUE op_asr(const VALUE other) const override;
            VALUE op_shr(const VALUE other) const override;
            VALUE op_and(const VALUE other) const override;
            VALUE op_xor(const VALUE other) const override;
            VALUE op_or(const VALUE other) const override;
    };
}


/* ***************************************************************************
* EVALUATOR
*/

namespace liteexpr {
    class Evaluator: public LiteExprBaseVisitor {
        SYMBOLS symbols;
        map<LiteExprParser::ExprContext*,VALUE> cache;

        public:
            Evaluator(SYMBOLS s);
            SYMBOLS getSymbols();

            any visitFile(LiteExprParser::FileContext *ctx) override;
            any visitString(LiteExprParser::StringContext *ctx) override;
            any visitDouble(LiteExprParser::DoubleContext *ctx) override;
            any visitHex(LiteExprParser::HexContext *ctx) override;
            any visitInt(LiteExprParser::IntContext *ctx) override;
            any visitArray(LiteExprParser::ArrayContext *ctx) override;
            any visitObject(LiteExprParser::ObjectContext *ctx) override;
            any visitPairlist(LiteExprParser::PairlistContext *ctx) override;
            any visitPair(LiteExprParser::PairContext *ctx) override;
            any visitList(LiteExprParser::ListContext *ctx) override;
            any visitCall(LiteExprParser::CallContext *ctx) override;
            any visitParen(LiteExprParser::ParenContext *ctx) override;
            any visitPrefixOp(LiteExprParser::PrefixOpContext *ctx) override;
            any visitPostfixOp(LiteExprParser::PostfixOpContext *ctx) override;
            any visitUnaryOp(LiteExprParser::UnaryOpContext *ctx) override;
            any visitBinaryOp(LiteExprParser::BinaryOpContext *ctx) override;
            any visitTernaryOp(LiteExprParser::TernaryOpContext *ctx) override;
            any visitAssignOp(LiteExprParser::AssignOpContext *ctx) override;
            any visitVariable(LiteExprParser::VariableContext *ctx) override;
            any visitMemberVar(LiteExprParser::MemberVarContext *ctx) override;
            any visitIndexedVar(LiteExprParser::IndexedVarContext *ctx) override;
            any visitSimpleVar(LiteExprParser::SimpleVarContext *ctx) override;
            any visitTerm(LiteExprParser::TermContext *ctx) override;
    };

    class Compiled {
        antlr4::ANTLRInputStream* input;
        antlr4::CommonTokenStream* tokens;
        antlr4::ParserRuleContext* parseTree;
        LiteExprLexer* lexer;
        LiteExprParser* parser;

        public:
            Compiled(string);
            ~Compiled();
            VALUE eval(SYMBOLS symbols);
    };

    Compiled compile(const string& expr);
    VALUE eval(const string& expr, SYMBOLS symbols);
}


/* ***************************************************************************
* EXCEPTIONS
*/

namespace liteexpr {
    class Error {
        string text;

        public:
            Error(const string& t);
            Error(const string& t, int line);
            Error(const string& t, int line, int col);
            operator string() const { return this->text; }
    };

    class SyntaxError: public Error {
        public:
            SyntaxError(const string& t, int line): Error(t, line) {};
            SyntaxError(const string& t, int line, int col): Error(t, line, col) {};
    };

    class RuntimeError: public Error {
        public:
            RuntimeError(const string& t): Error(t) {};
            RuntimeError(const string& t, int line): Error(t, line) {};
            RuntimeError(const string& t, int line, int col): Error(t, line, col) {};
    };

    class BasicSyntaxError: public Error {
        public:
            BasicSyntaxError(const string& t): Error(t) {};
    };

    class BasicRuntimeError: public Error {
        public:
            BasicRuntimeError(const string& t): Error(t) {};
    };
}


/* ***************************************************************************
* BUILTINS
*/

namespace liteexpr {
    extern map<string,VALUE> builtins;
}


/* ***************************************************************************
* HELPER FUNCTIONS
*/

namespace liteexpr {
    VALUE make_value(int);
    VALUE make_value(int64_t);
    VALUE make_value(double);
    VALUE make_value(const string &);
    VALUE make_value(initializer_list<VALUE>);
    VALUE make_value(initializer_list<pair<string,VALUE> >);
    INTEGER make_integer(int64_t);
    DOUBLE make_double(double);
    STRING make_string(const string&);
    ARRAY make_array(initializer_list<VALUE>);
    OBJECT make_object(initializer_list<pair<string,VALUE> >);
    SYMBOLS make_symbols(initializer_list<pair<string,VALUE> >);
}


#endif /* LITEEXPR_H_ */
