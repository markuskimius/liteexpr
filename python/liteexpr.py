import math
import antlr4
from LiteExprLexer import LiteExprLexer
from LiteExprParser import LiteExprParser
from LiteExprListener import LiteExprListener


##############################################################################
# CONSTANTS

INTMASK = 0xffffffffffffffff


##############################################################################
# PUBLIC FUNCTIONS

def evalfd(fd, symbols=None):
    code = fd.read()

    return eval(code, symbols)


def eval(code, symbols=None):
    compiled = compile(str(code))

    return compiled.eval(symbols)


def compile(code):
    try:
        lexer = LiteExprLexer(antlr4.InputStream(code))
        parser = LiteExprParser(antlr4.CommonTokenStream(lexer))
        parser._errHandler = antlr4.error.ErrorStrategy.BailErrorStrategy()
        tree = parser.file_()
    except antlr4.error.Errors.ParseCancellationException as e:
        t = e.args[0].offendingToken

        raise LE_SyntaxError(f"Unexpected token `{t.text}`", t.line, t.column) from None

    return LE_Compiled(tree)


def compilefd(fd):
    code = fd.read()

    return compile(code)


##############################################################################
# EXCEPTIONS

class LE_Error(Exception):
    def __init__(self, text, line=None, column=None):
        if   column is None and line is None : super().__init__(f"{text}")
        elif column is None                  : super().__init__(f"[line {line}] {text}")
        else                                 : super().__init__(f"[line {line}, col {column}] {text}")

class LE_SyntaxError(LE_Error): pass
class LE_RuntimeError(LE_Error): pass


##############################################################################
# HELPER CLASSES

class LE_Compiled:
    def __init__(self, tree):
        self.tree = tree
        self.walker = antlr4.ParseTreeWalker()

    def eval(self, symbolTable=None):
        evaluator = LE_Evaluator(symbolTable)

        self.walker.walk(evaluator, self.tree)

        return evaluator.result[self.tree]


class LE_SymbolTable:
    def __init__(self, symbols=dict(), parent=None):
        self.symbols = builtins | symbols if parent is None else symbols
        self.parent = parent

    def __getitem__(self, name):
        name = str(name)

        if name in self.symbols : return self.symbols[name]
        else                    : return self.parent[name]

    def __setitem__(self, name, value):
        name = str(name)

        self.symbols[name] = value

        return self.symbols[name]

    def __str__(self):
        import json

        return json.dumps(self.symbols, indent=2, default=str)


##############################################################################
# TYPES

class LE_Variable:
    def __init__(self, name, container):
        self.name = name
        self.container = container

    @property
    def value(self):
        if   isinstance(self.container, LE_SymbolTable) : return self.container[self.name]
        elif isinstance(self.container, list)           : return self.container[self.name]
        elif isinstance(self.container, dict)           : return self.container[self.name]
        else                                            : return self.container

    @value.setter
    def value(self, value):
        if   isinstance(self.container, LE_SymbolTable) : self.container[self.name] = value
        elif isinstance(self.container, list)           : self.container[self.name] = value
        elif isinstance(self.container, dict)           : self.container[self.name] = value
        else                                            : self.container = value

        return value


class LE_Int(int):
    @property
    def value(self):
        return self


class LE_Double(float):
    @property
    def value(self):
        return self


class LE_String(str):
    @property
    def value(self):
        return self


class LE_Array(list):
    @property
    def value(self):
        return self


class LE_Object(dict):
    @property
    def value(self):
        return self


##############################################################################
# EVALUATOR

class LE_Evaluator(LiteExprListener):
    def __init__(self, symbolTable=None):
        super().__init__()
        self.result = dict()
        self.symbolTable = LE_SymbolTable() if symbolTable is None else symbolTable

    def exitFile(self, ctx):
        lastExpr = ctx.expr()[-1]

        self.result[ctx] = self.result[lastExpr].value

    def exitString(self, ctx):
        try:
            self.result[ctx] = LE_String(_decodeString(ctx.STRING().getText()[1:-1]))
        except LE_SyntaxError as e:
            raise LE_SyntaxError(str(e), ctx.start.line, ctx.start.column) from None

    def exitDouble(self, ctx):
        self.result[ctx] = LE_Double(ctx.DOUBLE().getText())

    def exitHex(self, ctx):
        self.result[ctx] = LE_Int(ctx.HEX().getText()[2:], 16)

    def exitInt(self, ctx):
        self.result[ctx] = LE_Int(ctx.INT().getText())

    def exitCall(self, ctx):
        fn = self.result[ctx.varname()].value
        args = self.result[ctx.list_()].value

        try:
            self.result[ctx] = fn(self.symbolTable, *args)
        except LE_SyntaxError as e:
            raise LE_SyntaxError(f"Syntax error while executing {ctx.getText()}: {str(e)}", ctx.start.line, ctx.start.column) from None
        except LE_RuntimeError as e:
            raise LE_RuntimeError(f"Runtime error while executing {ctx.getText()}: {str(e)}", ctx.start.line, ctx.start.column) from None

    def exitVariable(self, ctx):
        self.result[ctx] = self.result[ctx.varname()]

    def exitObject(self, ctx):
        self.result[ctx] = self.result[ctx.pairlist()]

    def exitArray(self, ctx):
        self.result[ctx] = self.result[ctx.list_()]

    def exitParen(self, ctx):
        self.result[ctx] = self.result[ctx.expr()]

    def exitPostfixOp(self, ctx):
        var = self.result[ctx.varname()]
        op = ctx.op.text
        T = type(var.value)

        if   op == "++" : self.result[ctx] = var.value; var.value = T(var.value + 1)
        elif op == "--" : self.result[ctx] = var.value; var.value = T(var.value - 1)
        else            : raise LE_SyntaxError("Unknown postfix operator `{op}`", ctx.start.line, ctx.start.column)

    def exitPrefixOp(self, ctx):
        var = self.result[ctx.varname()]
        op = ctx.op.text
        T = type(var.value)

        if   op == "++" : var = T(var.value + 1); self.result[ctx] = var.value
        elif op == "--" : var = T(var.value - 1); self.result[ctx] = var.value
        else            : raise LE_SyntaxError("Unknown prefix operator `{op}`", ctx.start.line, ctx.start.column)

    def exitUnaryOp(self, ctx):
        value = self.result[ctx.expr()].value
        op = ctx.op.text
        T = type(value)

        if   op == "!"  : self.result[ctx] = LE_Int(not value)
        elif op == "~"  : self.result[ctx] = LE_Int(~value)
        elif op == "+"  : self.result[ctx] = value
        elif op == "-"  : self.result[ctx] = T(-value)
        else            : raise LE_SyntaxError("Unknown unary operator `{op}`", ctx.start.line, ctx.start.column)

    def exitBinaryOp(self, ctx):
        op = ctx.op.text
        value = (
            self.result[ctx.expr(0)].value,
            self.result[ctx.expr(1)].value,
        )
        T = LE_Int if isinstance(value[0],int) and isinstance(value[1],int) else LE_Double

        if   op == "**"  : self.result[ctx] = (LE_Double if value[1] < 0 else T)(value[0] **  value[1])
        elif op == "*"   : self.result[ctx] = T(value[0] *   value[1])
        elif op == "/"   : self.result[ctx] = T(value[0] /   value[1])
        elif op == "%"   : self.result[ctx] = LE_Int(value[0] %   value[1])
        elif op == "+"   : self.result[ctx] = LE_String(str(value[0]) + str(value[1])) if isinstance(value[0],str) or isinstance(value[1],str) else T(value[0] + value[1])
        elif op == "-"   : self.result[ctx] = T(value[0] -   value[1])
        elif op == "<<"  : self.result[ctx] = LE_Int(value[0] <<  value[1])
        elif op == ">>"  : self.result[ctx] = LE_Int(value[0] >>  value[1])
        elif op == ">>>" : self.result[ctx] = LE_Int((value[0] & INTMASK) >> value[1])
        elif op == "<"   : self.result[ctx] = LE_Int(value[0] <   value[1])
        elif op == "<="  : self.result[ctx] = LE_Int(value[0] <=  value[1])
        elif op == ">"   : self.result[ctx] = LE_Int(value[0] >   value[1])
        elif op == ">="  : self.result[ctx] = LE_Int(value[0] >=  value[1])
        elif op == "=="  : self.result[ctx] = LE_Int(value[0] ==  value[1])
        elif op == "!="  : self.result[ctx] = LE_Int(value[0] !=  value[1])
        elif op == "&"   : self.result[ctx] = LE_Int(value[0] &   value[1])
        elif op == "^"   : self.result[ctx] = LE_Int(value[0] ^   value[1])
        elif op == "|"   : self.result[ctx] = LE_Int(value[0] |   value[1])
        elif op == "&&"  : self.result[ctx] = LE_Int(value[0] and value[1])
        elif op == "||"  : self.result[ctx] = LE_Int(value[0] or  value[1])
        else             : raise LE_SyntaxError("Unknown binary operator `{op}`", ctx.start.line, ctx.start.column)

    def exitAssignOp(self, ctx):
        op = ctx.op.text
        var = self.result[ctx.varname()]
        value2 = self.result[ctx.expr()].value

        if op == "=" : self.result[ctx] = value2
        else         :
            T = LE_Int if isinstance(var.value,int) and isinstance(var.value,int) else LE_Double

            if   op == "**="  : self.result[ctx] = (LE_Double if value2 < 0 else T)(var.value **  value2)
            elif op == "*="   : self.result[ctx] = T(var.value *   value2)
            elif op == "/="   : self.result[ctx] = T(var.value /   value2)
            elif op == "%="   : self.result[ctx] = LE_Int(var.value %   value2)
            elif op == "+="   : self.result[ctx] = LE_String(str(var.value) + str(value2)) if isinstance(var.value,str) or isinstance(value2,str) else T(var.value + value2)
            elif op == "-="   : self.result[ctx] = T(var.value -   value2)
            elif op == "<<="  : self.result[ctx] = LE_Int(var.value <<  value2)
            elif op == ">>="  : self.result[ctx] = LE_Int(var.value >>  value2)
            elif op == ">>>=" : self.result[ctx] = LE_Int((var.value & INTMASK) >> value2)
            elif op == "&="   : self.result[ctx] = LE_Int(var.value &   value2)
            elif op == "^="   : self.result[ctx] = LE_Int(var.value ^   value2)
            elif op == "|="   : self.result[ctx] = LE_Int(var.value |   value2)
            elif op == "&&="  : self.result[ctx] = LE_Int(var.value and value2)
            elif op == "||="  : self.result[ctx] = LE_Int(var.value or  value2)
            else              : raise LE_SyntaxError("Unknown assign operator `{op}`", ctx.start.line, ctx.start.column)

        var.value = self.result[ctx]

    def exitTertiaryOp(self, ctx):
        op = (
            ctx.op1.text,
            ctx.op2.text,
        )
        value = (
            self.result[ctx.expr(0)].value,
            self.result[ctx.expr(1)].value,
            self.result[ctx.expr(2)].value,
        )

        if op[0] == "?" and op[1] == ":" : self.result[ctx] = LiteExprDatum(value[1] if value[0] else value[2])
        else                             : raise LE_SyntaxError("Unknown tertiary operator `{op[0]} {op[1]}`", ctx.start.line, ctx.start.column)

    def exitIndexedVar(self, ctx):
        array = self.result[ctx.varname()].value
        index = self.result[ctx.expr()].value

        self.result[ctx] = LE_Variable(index, array)

    def exitMemberVar(self, ctx):
        base = self.result[ctx.varname(0)].value
        member = self.result[ctx.varname(1)].name

        self.result[ctx] = LE_Variable(member, base)

    def exitSimpleVar(self, ctx):
        varname = ctx.ID().getText()

        self.result[ctx] = LE_Variable(varname, self.symbolTable)

    def exitPairlist(self, ctx):
        self.result[ctx] = LE_Object()

        for pair in ctx.pair():
            for k,v in self.result[pair].items():
                self.result[ctx][k] = v

    def exitPair(self, ctx):
        name = ctx.ID().getText()
        value = self.result[ctx.expr()]

        self.result[ctx] = { LE_String(name) : value }

    def exitList(self, ctx):
        self.result[ctx] = LE_Array()

        for item in ctx.expr():
            self.result[ctx] += [self.result[item].value]


##############################################################################
# HELPER FUNCTIONS

def _decodeString(s):
    decoded = ""
    i = 0

    while i < len(s):
        if   s[i:i+2] == "\\\\"   : decoded += "\\"; i+=2
        elif s[i:i+2] == "\\\""   : decoded += "\""; i+=2
        elif s[i:i+3] == "\\\r\n" : i+=2
        elif s[i:i+2] == "\\\r"   : i+=2
        elif s[i:i+2] == "\\\n"   : i+=2
        elif s[i:i+2] == "\\t"    : decoded += "\t"; i+=2
        elif s[i:i+2] == "\\r"    : decoded += "\r"; i+=2
        elif s[i:i+2] == "\\n"    : decoded += "\n"; i+=2
        elif s[i:i+2] == "\\x"    : decoded += chr(int(s[i+2:i+4], 16)); i+=4
        elif s[i:i+2] == "\\u"    : decoded += chr(int(s[i+2:i+6], 16)); i+=6
        elif s[i:i+2] == "\\U"    : decoded += chr(int(s[i+2:i+10], 16)); i+=10
        elif s[i:i+1] == "\\"     : raise LE_SyntaxError(f"Invalid backslash sequence in string at position {i}")
        else                      : decoded += s[i]; i+=1

    return decoded


def _encodeString(s):
    encoded = ""

    for c in s:
        if   c == "\\" : encoded += "\\\\"
        elif c == "\"" : encoded += "\\\""
        else           : encoded += c

    return encoded


##############################################################################
# BUILTIN FUNCTIONS

builtins = {
    "CEIL"  : lambda s,  x : LE_Int(math.ceil(x)),
    "EVAL"  : lambda s,  x : eval(x, s),
    "FLOOR" : lambda s,  x : LE_Int(math.floor(x)),
    "FOR"   : lambda s,  a, b, c, d : __builtin_for(s, a, b, c, d),
    "IF"    : lambda s, *x : __builtin_if(s, *x),
    "LEN"   : lambda s,  x : LE_Int(len(x)),
    "PRINT" : lambda s, *x : __builtin_print(s, *x),
    "ROUND" : lambda s, *x : LE_Int(round(*x)),
    "SQRT"  : lambda s,  x : LE_Double(math.sqrt(x)),
    "WHILE" : lambda s,  a, b: __builtin_while(s, a, b),
}

def __builtin_for(s, a, b, c, d):
    eval(a, s)

    while(eval(b, s)):
        eval(d, s)
        eval(c, s)

    return LE_Int(0)

def __builtin_if(s, *x):
    v = x[-1]
    i = 0

    while i+1 < len(x):
        if eval(x[i], s): v = x[i+1]
        i += 2

    return eval(v, s)

def __builtin_print(s, *x):
    o = []

    for y in x:
        o += [y.value]

    print(*o)

    return LE_Int(0)

def __builtin_while(s, a, b):
    while(eval(a, s)):
        eval(b, s)

    return LE_Int(0)

