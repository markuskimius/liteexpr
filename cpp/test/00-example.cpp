#include <string>
#include <iostream>
#include "liteexpr.h"

int main(int argc, const char* argv[]) {
    liteexpr::SYMBOLS symbols = liteexpr::make_symbols({
        { "grades", liteexpr::make_value({
            { "alice", liteexpr::make_value("A") },
            { "bob"  , liteexpr::make_value("B") },
        })},
    });

    try {
        liteexpr::eval(R"(
            PRINT("I have " + LEN(grades) + " students");
            PRINT("Alice's grade is " + grades.alice);
        )", symbols);
    }
    catch(liteexpr::Error e) {
        std::cerr << std::string(e) << std::endl;
        return 1;
    }

    return 0;
}
