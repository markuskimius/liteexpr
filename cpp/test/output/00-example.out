I have 2 students
Alice's grade is A
